src/parse_args.cpp

```diff
@@ -0,0 +1,111 @@
+#include "parse_args.hpp"
+#include "util/expected.hpp"
+#include <algorithm>
+#include <regex>
+#include <set>
+#include <stdexcept>
+
+#include "config.hpp"
+#include <slurm/spank.h>
+
+#define LINUX_FPATH "[^\\0,:]+"
+
+namespace impl {
+
+const std::regex protocol_pattern("^(jfrog|file|https):\\/\\/"
+                                  "(" LINUX_FPATH ")"
+                                  "(:" LINUX_FPATH ")?",
+                                  std::regex::ECMAScript);
+const std::regex default_pattern("(" LINUX_FPATH ")"
+                                 "(:" LINUX_FPATH ")?",
+                                 std::regex::ECMAScript);
+
+const std::map<std::string, enum protocol> protocol_from_string {
+  {"file", protocol::file},
+  {"https", protocol::https},
+  {"jfrog", protocol::jfrog}
+};
+
+std::vector<std::string> split(const std::string &s, char delim) {
+  std::vector<std::string> elems;
+  std::stringstream ss(s);
+  std::string item;
+  while (std::getline(ss, item, delim)) {
+    if (!item.empty())
+      elems.push_back(item);
+  }
+  return elems;
+}
+
+util::expected<std::vector<mount_entry>, std::runtime_error>
+parse_arg(const std::string &arg) {
+  std::vector<mount_entry> entries;
+  std::vector<std::string> arguments = split(arg, ',');
+
+  for (auto &entry : arguments) {
+    std::smatch match_pieces;
+    if (std::regex_match(entry, match_pieces, protocol_pattern)) {
+      // [protocol]::/[path]:[mountpoint]
```

Comment:
I'm not sure we agreed that '/' has to precede the path. Otherwise the user couldn't load their own images.

Reply:
There is a problem in how slurm processes / inherits the flags. There is no way (at least I haven't found one) to determine if `--uenv` has been set in the current job step (srun) or if it has been inherited from salloc/sbatch. Since the user might have changed the directory in the meantime, it's not possible to unambiguously resolve relative paths. 

I think we should be able to fix this with the shell function we planned for `uenv` command:

```
#SBATCH ...

# -> sets only UENV_MOUNT_LIST environment variable  when run without CMD arg
uenv mount|activate image:mountpoint

# possibly change directory 

srun cmd

```